File: Leetcode/tree/1110_DeleteNodesAndReturnForest/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        List<TreeNode> res = new ArrayList<>();
        dfs_mark(root, to_delete);
        if(root.val!=-1) res.add(root);
        dfs_add(root, res);
        dfs_delete(root);
        return res;
    }
        private void dfs_mark(TreeNode root, int[] to_delete){
            if(root==null) return;
            for(int num: to_delete){
                    if(root.val == num){
                        root.val = -1;
                    }
                }
                dfs_mark(root.left, to_delete);
                dfs_mark(root.right, to_delete);
            }
        private void dfs_add(TreeNode root, List<TreeNode> res){
            if(root==null) return;
            if(root.val == -1){
                if(root.left!=null && root.left.val!=-1)
                    res.add(root.left);
                if(root.right!=null && root.right.val!=-1)
                    res.add(root.right);
            }
            dfs_add(root.left, res);
            dfs_add(root.right, res);
        }
        private TreeNode dfs_delete(TreeNode root){
            if(root==null) return null;
            root.left = dfs_delete(root.left);
            root.right = dfs_delete(root.right);
            if(root.val == -1) return null;
            return root;
        }
}

File: Leetcode/tree/1110_DeleteNodesAndReturnForest/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def delNodes(self, root, to_delete):
        """
        :type root: TreeNode
        :type to_delete: List[int]
        :rtype: List[TreeNode]
        """

        to_delete_set = set(to_delete)
        res = []
        def dfs(root, is_root):
            if not root: return None
            deleted = root.val in to_delete_set
            if is_root and not deleted:
                res.append(root)
            root.left = dfs(root.left, deleted)
            root.right = dfs(root.right, deleted)
            return None if deleted else root
        dfs(root, True)
        return res
